src/software/thunderscope/thunderscope_main.py

```diff
@@ -192,54 +198,74 @@
+    #
+    #       Run Blue
+    #       Run Yellow
+    #       Run Blue + Diagnostics
+    #       Run Yellow + Diagnostics
+    #       Run Diagnostics
+    #
+    # We want to run either 1 instance of AI or 1 instance of RobotCommunication or both which will
+    # send/recv packets over the provided multicast channel.
 
+    if args.run_blue or args.run_yellow or args.run_diagnostics:
         tscope = Thunderscope(
             layout_path=args.layout,
-            load_blue=False,
-            load_yellow=True,
-            load_diagnostics=True,
+            load_blue=bool(args.run_blue),
+            load_yellow=bool(args.run_yellow),
+            load_diagnostics=bool(args.run_diagnostics),
             load_gamecontroller=False,
             visualization_buffer_size=args.visualization_buffer_size,
         )
 
-        proto_unix_io = tscope.yellow_full_system_proto_unix_io
-        runtime_dir = args.yellow_full_system_runtime_dir
-        friendly_colour_yellow = True
-        debug = args.debug_yellow_full_system
+        full_system_proto_unix_io = None
+        diagnostics_proto_unix_io = None
+
+        if args.run_blue:
+            full_system_proto_unix_io = tscope.blue_full_system_proto_unix_io
+            runtime_dir = args.blue_full_system_runtime_dir
+            friendly_colour_yellow = False
+            debug = args.debug_blue_full_system
+        elif args.run_yellow:
+            full_system_proto_unix_io = tscope.yellow_full_system_proto_unix_io
+            runtime_dir = args.yellow_full_system_runtime_dir
+            friendly_colour_yellow = True
+            debug = args.debug_yellow_full_system
+
+        if args.run_diagnostics:
+            diagnostics_proto_unix_io = tscope.robot_diagnostics_proto_unix_io
+
+        if args.run_blue or args.run_yellow:
+            with ProtoLogger(
+                args.blue_full_system_runtime_dir,
+            ) as blue_logger, ProtoLogger(
+                args.yellow_full_system_runtime_dir,
+            ) as yellow_logger, FullSystem(
+                runtime_dir, debug, friendly_colour_yellow
+            ) as full_system:
+
+                print(type(blue_logger))
+
+                full_system_proto_unix_io.register_to_observe_everything(
+                    blue_logger.buffer
+                )
+                full_system_proto_unix_io.register_to_observe_everything(
+                    yellow_logger.buffer
+                )
+                full_system.setup_proto_unix_io(full_system_proto_unix_io)
+
+        with RobotCommunication(
+            full_system_proto_unix_io,
+            diagnostics_proto_unix_io,
+            getRobotMulticastChannel(0),
+            args.interface,
+        ) as robot_communication:
+            if args.run_diagnostics:
+                tscope.toggle_robot_connection_signal.connect(
+                    robot_communication.toggle_robot_connection
+                )
 
```

Comment:
`with` statements in Python are used as context managers. When you enter the block, the `__enter__` method of the class is called, and once finished with the `with` block, the `__exit__` statement is called. So since you have two `with` statements in this code, the `ProtoLogger` and the `FullSystem` are actually deleted (`__exit__` called) before you reach the second `with` statement. I'm not very familiar with the different ways of using the `with` statement, but I think you need to combine your logic to be all with in the same `with` block so it properly enters and exists.